src/room_buffer.rs

```diff
@@ -196,11 +204,89 @@ impl RoomBuffer {
         }
     }
 
-    pub fn update_buffer_name(&mut self) {
+    pub fn update_buffer_name(&self) {
         let name = self.calculate_buffer_name();
         self.weechat_buffer().set_name(&name)
     }
 
+    /// Send out a typing notice.
+    ///
+    /// This will send out a typing notice or reset one if needed. It will make
+    /// sure that only one typing notice request is in flight at a time.
+    ///
+    /// Typing notices are sent out only if we have more than 4 letters in the
+    /// input and the input isn't a command.
+    ///
+    /// If the input is empty the typing notice is disabled.
+    pub fn update_typing_notice(&self) {
+        // We're in the process of sending out a typing notice, so don't make
+        // the same request twice.
+        if self.typing_in_flight.load(Ordering::Relaxed) {
+            return;
+        }
+
+        let buffer = self.weechat_buffer();
+        let input = buffer.input();
+
+        if input.starts_with("/") && !input.starts_with("//") {
+            // Don't send typing notices for commands.
+            return;
+        }
+
+        let connection = self.inner.connection.clone();
+        let room_id = self.room_id.clone();
+        let user_id = self.own_user_id.clone();
+        let typing_notice_time = self.inner.typing_notice_time.clone();
+        let typing_in_flight = self.typing_in_flight.clone();
+
+        let send = async move |typing: bool| {
+            let typing_time = typing_notice_time;
+
+            if let Some(connection) = &*connection.borrow() {
+                let response = connection
+                    .send_typing_notice(&*room_id, &*user_id, typing)
+                    .await;
+
+                // If we got a successfull response mark when we sent out the
+                // notice so we don't send out another one while this one is
+                // active, or if we disabled the notice mark that no notice is
+                // active.
+                if response.is_ok() {
+                    if typing {
+                        *typing_time.borrow_mut() = Some(Instant::now());
+                    } else {
+                        *typing_time.borrow_mut() = None;
+                    }
+                }
+            };
+
+            typing_in_flight.store(false, Ordering::Relaxed);
+        };
+
+        let typing_time = self.inner.typing_notice_time.borrow_mut();
+
+        if input.len() < 4 && typing_time.is_some() {
+            // If we have an active typing notice and our input is short, e.g.
+            // we removed the input set the typing notice to false.
+            self.typing_in_flight.store(true, Ordering::Release);
```

Comment:
Can this be moved inside the `send` future?

Reply:
Yes but you'll end up spawning futures even when you don't need them, also there's no guarantee that our future will run before the next time the input_changed signal is fired.

Reply:
> Yes but you'll end up spawning futures even when you don't need them

This can happen now too, no? In cases where there's a race between the above `store` and the next call of `update_typing_notice`.

But I agree, the future is not guaranteed to run immediately so that exacerbates it, let's leave it how it is.

Reply:
There's no race here, another `update_typing_notice()` can't be called before we yield control back to Weechat, we do that once we return from this method. The future on the other hand can run in any order Weechat decides to run it.

Reply:
Ah, right, it's not async itself.

Reply:
Even if it was async, unless there's an `await`, a yield point, Weechat won't run another callback since it's single threaded.

Reply:
Of course. I think Weechat's "signal" nomenclature got to me so I was applying an incorrect mental model.